### PR TITLE
Don't throw exceptions from ~_single_instance_hook

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Resolves: #2511 

## Description

C++ destructors are noexcept by definition. Throwing from a destructor is a noexcept violation which results in abort. Avoid this by making best effort to cleanup in destructors.

## Testing

CI/CD

## Documentation

No.

### CODE REVIEW

**Review by dv-msft — APPROVED** [review]

File: tests/end_to_end/helpers.h [hunk 1]
```diff
@@ -129,7 +129,11 @@ typedef class _single_instance_hook : public _hook_helper
     }
     ~_single_instance_hook()
     {
-        detach();
+        // Best effort cleanup. Ignore errors.
+        if (link_object) {
+            (void)ebpf_link_detach(link_object);
```
Comment (saxena-anurag) [inline tests/end_to_end/helpers.h:134]: Though unrelated to changes in this PR, it looks like `detach()` has a bug: It is throwing an exception when `ebpf_link_detach()` returns EBPF_SUCCESS.

```    void
    detach()
    {
        if (link_object != nullptr) {
            if (ebpf_link_detach(link_object) == EBPF_SUCCESS) {
                throw std::runtime_error("ebpf_link_detach failed");
            }
            ebpf_link_close(link_object);
            link_object = nullptr;
        }
    }
```

**Review by saxena-anurag — APPROVED** [review]
